src/function.jl

```diff
@@ -1,15 +1,72 @@
 abstract type AbstractOptimizationFunction end
+abstract type AbstractADType end
 
-struct OptimizationFunction{F,G,H,K} <: AbstractOptimizationFunction
+struct AutoForwardDiff <: AbstractADType end
+struct AutoReverseDiff <: AbstractADType end
+struct AutoTracker <: AbstractADType end
+struct AutoZygote <: AbstractADType end
+struct AutoFiniteDiff <: AbstractADType end
+struct AutoModelingToolkit <: AbstractADType end
+
+struct OptimizationFunction{F,G,H,HV,K} <: AbstractOptimizationFunction
     f::F
     grad::G
     hess::H
+    hv::HV
+    adtype::AbstractADType
     kwargs::K
 end
 
-function OptimizationFunction(f,x; grad=nothing,hess=nothing, p=DiffEqBase.NullParameters(),kwargs...)
-    _f = θ -> f(θ,p)
-    g! = (res,θ) -> ForwardDiff.gradient!(res, _f, θ)
-    h! = (res,θ) -> ForwardDiff.hessian!(res, _f, θ)
-    return OptimizationFunction{typeof(f),typeof(g!),typeof(h!),typeof(kwargs)}(f,g!,h!,kwargs)
+function OptimizationFunction(f, x, ::AutoForwardDiff; grad=nothing,hess=nothing, p=DiffEqBase.NullParameters(), chunksize = 1, hv = nothing, kwargs...)
```

Comment:
`p` probably shouldn't be here.

Reply:
We would need it for https://github.com/SciML/GalacticOptim.jl/pull/17/files#diff-1fee433add706d5d530efb579f0d3fbaR21?